### PR TITLE
implement streaming data-loader for log anomaly detection model training to reduce memory pressure

### PR DESCRIPTION
**Pull request by tybalex**
details in OEP: https://github.com/rancher/opni/pull/1026
issue: https://github.com/rancher/opni/issues/1095

** what changes in this PR: **
1. add class `IterablePaddedDataset` that defines how to collect data for model training.
2. add `get_streaming_dataloader()` that returns a streaming dataloader which makes use of IterablePaddedDataset. By default, num_workers = 3
3. add functions `get_padded_data()` and `prepare_metadata()` to pad and mask data from streaming dataloader and  feed into model.
4. add `yield_all_training_data()` and `mask_batch()` that downloads and preprocess training data.
5. add `get_weights()` to apply weighted random sampling to training data.
6. Adjust logic in `opni_trainer.py` to incorporate all the changes above
7. other minor fixes, includes fix training time in streaming mode, etc..

### CODE REVIEW

File: opni_inference_service/models/opnilog/opnilog_model.py [hunk 1]
```diff
@@ -284,17 +285,39 @@ def make_std_mask(tgt, pad):
...
     ):
         self.c = copy.deepcopy
 
         self.data = data
         self.padded_data = self._get_padded_data(data, pad_len)
         self.mask_percentage = mask_percentage
-        self.transforms = transforms
-        self.pad = pad
+        # self.transforms = transforms
+        self.transforms = None
```
Comment (AmartC) [inline opni_inference_service/models/opnilog/opnilog_model.py:320]: Any reason why this isn't being set to the transforms variable?

**Review by AmartC — DISMISSED** [review]

**Review by AmartC — APPROVED** [review]
